sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/FlatteningDeserializer.java

```diff
@@ -65,58 +68,185 @@ public static SimpleModule getModule(final ObjectMapper mapper) {
         SimpleModule module = new SimpleModule();
         module.setDeserializerModifier(new BeanDeserializerModifier() {
             @Override
-            public JsonDeserializer<?> modifyDeserializer(DeserializationConfig config, BeanDescription beanDesc,
+            public JsonDeserializer<?> modifyDeserializer(DeserializationConfig config,
+                                                          BeanDescription beanDesc,
                                                           JsonDeserializer<?> deserializer) {
-                if (beanDesc.getBeanClass().getAnnotation(JsonFlatten.class) != null) {
+                if (BeanDeserializer.class.isAssignableFrom(deserializer.getClass())) {
```

Comment:
I am not sure if opening it up is good or bad at this point.

Here is a bug in track1 that kind of *caused* by open-up if there is any bug in Flattening (not really, but previous class without "@JsonFlatten" will not be affected, but now it will)
https://github.com/Azure/autorest-clientruntime-for-java/issues/663

And we got this one there
https://github.com/Azure/azure-sdk-for-java/issues/8372

Just saying it could be risky if there is unfixed or even unknown bug.

Reply:
I think we're safe to do this in "deserializer" because in deserialization path we always check for `JsonProperty` value and if it has un-escaped dots then it means "do flatten", right?

\\cc @jianghaolu - please chime if this interpretation in "deserialization" context is wrong

Reply:
Sorry, I think I mean the same place in FlatteningSerializer (it is opened-up as well). Both bug referred is happening in serialization stage.

Reply:
Right, this PR is to address the discriminated type "deserialization" issue, https://github.com/Azure/azure-sdk-for-java/issues/10217.

Thanks for the heads-up on the "serialization" path. Reverted the  FlatteningSerializer, so there won't be any change in the `FlatteningSerializer` and "serialization" logic stay as it is. This means the track1 issue with template "serialization" "663" will not happen when using azure-core. The "8372" is a different issue, and this PR does not address it.

Reply:
@weidongxu-microsoft  - [reg](https://github.com/Azure/azure-sdk-for-java/pull/10310#discussion_r409970231) -  I had a chat with Jinaghao on AutoRest logic associated with escaping dots in `JsonProperty`. Such escaping will be done only if the properties in the parent type have to include non-escaped dots, this means we should limit the application of Deserializer only to JsonFlatten annotated types.

Reply:
@anuchandy Got it, as well as the required "@JsonFlatten" on there classes as well.

One question, at least for the ARM template (which should not have any flatten in there), should we always use `readValue` to convert json to an `Object` (as `LinkedHashMap`), instead of using `readTree` to `ObjectNode`, so that if FlattenSerializer got involved, it could still escape the "." as it is now a `Map` with nested `Map`s?